### PR TITLE
Replace print statements with logger

### PR DESCRIPTION
**Pull request by botanical**
Addresses https://github.com/nsidc/earthaccess/issues/511

----
📚 Documentation preview 📚: https://earthaccess--566.org.readthedocs.build/en/566/

### CODE REVIEW

**Comment by botanical** [comment]
I didn't update the notebooks which also contain `print()` statements. Should I update those as well? 🤔

**Comment by mfisher87** [comment]
Good question! I think the notebooks should probably stay the same.

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,7 @@
 * Bug fixes:
     * fixed 483 by extracting a common CMR query method for collections and granules using SearchAfter header
     * Added VCR support for verifying the API call to CMR and the parsing of returned results without relying on CMR availability post development
+    * Updated to use logging library in place of print statements throught the repo.
```
Comment (mfisher87) [inline CHANGELOG.md:9]:
```suggestion
    * Updated to use logging library in place of print statements throughout the repo.
```

File: earthaccess/api.py [hunk 2]
```diff
@@ -64,7 +68,7 @@ def search_datasets(count: int = -1, **kwargs: Any) -> List[DataCollection]:
         ```
     """
     if not validate.valid_dataset_parameters(**kwargs):
-        print(
+        logger.warn(
             "Warning: a valid set of parameters is needed to search for datasets on CMR"
```
Comment (mfisher87) [inline earthaccess/api.py:72]:
```suggestion
            "A valid set of parameters is needed to search for datasets on CMR"
```

File: earthaccess/store.py [hunk 3]
```diff
@@ -110,7 +113,7 @@ def __init__(self, auth: Any, pre_authorize: bool = False) -> None:
                     self.set_requests_session(url)
 
         else:
-            print("Warning: the current session is not authenticated with NASA")
+            logger.warn("Warning: the current session is not authenticated with NASA")
```
Comment (mfisher87) [inline earthaccess/store.py:116]:
```suggestion
            logger.warn("The current session is not authenticated with NASA")
```

**Review by mfisher87 — COMMENTED** [review]
Looks wonderful, thank you for taking this on! :star_struck: 

A couple nitpicks :) What do you think?

**Comment by mfisher87** [comment]
Oo, also, what do you think of enabling ruff ruleset [T20](https://docs.astral.sh/ruff/rules/#flake8-print-t20) to prevent us from re-introducing print statements?

**Review by chuckwondo — CHANGES_REQUESTED** [review]
Hi @botanical! Glad to see your PR! Just a couple of tweaks, please, but otherwise, great start to your getting involved here!

**Comment by chuckwondo** [comment]
> Oo, also, what do you think of enabling ruff ruleset [T20](https://docs.astral.sh/ruff/rules/#flake8-print-t20) to prevent us from re-introducing print statements?

Good call. We would just want to disable this check for the lines in `print_help`, where we actually do want `print`, although I'd like to simply remove that method. It seems superfluous, and is not called internally, but perhaps that's a separate issue.

File: earthaccess/search.py [hunk 4]
```diff
@@ -306,8 +309,8 @@ def parameters(self, **kwargs: Any) -> Self:
 
     def print_help(self, method: str = "fields") -> None:
         """Prints the help information for a given method."""
-        print("Class components: \n")
-        print([method for method in dir(self) if method.startswith("_") is False])
+        logger.info("Class components: \n")
+        logger.info([method for method in dir(self) if method.startswith("_") is False])
```
[the remark this answers is not in the document]
Comment (mfisher87) [reply earthaccess/search.py:313]: Good catch!

**Comment by chuckwondo** [comment]
> Oo, also, what do you think of enabling ruff ruleset [T20](https://docs.astral.sh/ruff/rules/#flake8-print-t20) to prevent us from re-introducing print statements?

@botanical, if you can add this to the PR, that would be great. Note that at a minimum you'll have to pull in the latest changes from `main` to resolve conflicts in `CHANGELOG.md`. I recommend pulling the latest from upstream main and rebasing your branch on that, but that's not a requirement.

File: earthaccess/search.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (botanical) [reply earthaccess/search.py]: https://github.com/nsidc/earthaccess/pull/566/commits/3fadd28f4fa0ad1facd4484985fe0a76e8854e36 fixed!

File: earthaccess/api.py [hunk 5]
```diff
@@ -199,8 +203,8 @@ def download(
     try:
         results = earthaccess.__store__.get(granules, local_path, provider, threads)
     except AttributeError as err:
-        print(err)
-        print("You must call earthaccess.login() before you can download data")
+        logger.error(err)
+        logger.info("You must call earthaccess.login() before you can download data")
```
Comment (chuckwondo) [inline earthaccess/api.py:207]:
```suggestion
        logger.error(f"{err}: You must call earthaccess.login() before you can download data")
```

File: earthaccess/auth.py [hunk 6]
```diff
@@ -222,10 +222,10 @@ def get_s3_credentials(
                         cumulus_resp.url, allow_redirects=True, timeout=15
                     )
                     if not (auth_resp.ok):
-                        print(
+                        logger.error(
                             f"Authentication with Earthdata Login failed with:\n{auth_resp.text[0:1000]}"
                         )
-                        print(
+                        logger.info(
```
Comment (chuckwondo) [inline earthaccess/auth.py:228]:
```suggestion
                        logger.error(
```

File: earthaccess/store.py [hunk 7]
```diff
@@ -612,11 +615,11 @@ def _download_file(self, url: str, directory: Path) -> str:
                         # https://docs.python-requests.org/en/latest/user/quickstart/#raw-response-content
                         shutil.copyfileobj(r.raw, f, length=1024 * 1024)
             except Exception:
-                print(f"Error while downloading the file {local_filename}")
-                print(traceback.format_exc())
+                logger.error(f"Error while downloading the file {local_filename}")
+                logger.error(traceback.format_exc())
```
Comment (chuckwondo) [inline earthaccess/store.py:619]: Using `logger.exception` will include the traceback, so there's no need for an additional call.

```suggestion
                logger.exception(f"Error while downloading the file {local_filename}")
```

File: earthaccess/store.py [hunk 8]
```diff
@@ -660,7 +663,7 @@ def _open_urls_https(
             try:
                 fileset = _open_files(url_mapping, https_fs, threads)
             except Exception:
-                print(
+                logger.error(
                     "An exception occurred while trying to access remote files via HTTPS: "
                     f"Exception: {traceback.format_exc()}"
                 )
```
Comment (chuckwondo) [inline earthaccess/store.py:669]:
```suggestion
                logger.exception(
                    "An exception occurred while trying to access remote files via HTTPS"
                )
```

File: tests/integration/test_auth.py [hunk 9]
```diff
@@ -62,7 +62,7 @@ def test_auth_throws_exception_if_netrc_is_not_present():
     with pytest.raises(Exception) as e_info:
         earthaccess.login(strategy="netrc")
         assertions.assertRaises(FileNotFoundError)
-        print(e_info)
+        logger.info(e_info)
```
Comment (chuckwondo) [inline tests/integration/test_auth.py:65]: This is a test function, so we don't need to log a message.

```suggestion
```

File: tests/integration/test_auth.py [hunk 10]
```diff
@@ -87,12 +87,12 @@ def test_auth_can_fetch_s3_credentials():
...
-                print(f"Testing S3 credentials for {daac['short-name']}")
+                logger.info(f"Testing S3 credentials for {daac['short-name']}")
                 credentials = earthaccess.get_s3_credentials(daac["short-name"])
                 assertions.assertIsInstance(credentials, dict)
                 assertions.assertTrue("accessKeyId" in credentials)
             except Exception as e:
-                print(
+                logger.error(
                     f"An error occured while trying to fetch S3 credentials for {daac['short-name']}: {e}"
                 )
```
Comment (chuckwondo) [inline tests/integration/test_auth.py:97]: We shouldn't be logging messages from test functions. In fact, I suspect the `try`/`except` should be removed. If an exception occurs during the test, the test should fail. However, I don't know if this is an exception to that rule (no pun intended).

File: tests/unit/test_auth.py [hunk 11]
```diff
@@ -107,7 +110,7 @@ def test_auth_fails_for_wrong_credentials(self, user_input, user_password):
         auth = Auth()
         auth.login(strategy="interactive")
         with pytest.raises(Exception) as e_info:
-            print(e_info)
+            logger.info(e_info)
```
Comment (chuckwondo) [inline tests/unit/test_auth.py:113]:
```suggestion
```

**Review by chuckwondo — CHANGES_REQUESTED** [review]
@botanical, sorry, I made a number of individual comments rather than unified review comments.

File: tests/integration/test_auth.py (hunk 10 above)
Comment (botanical) [reply tests/integration/test_auth.py:96]: https://github.com/botanical/earthaccess/commit/64ec1355f707f1c457415d723f4b43ddaf4ba456 it seems intentional to me according to this commit 😯

File: pyproject.toml [hunk 12]
```diff
@@ -127,6 +127,9 @@ src = ["earthaccess", "stubs", "tests"]
 [tool.ruff.lint]
 extend-select = ["I", "T20"]
 
+[tool.ruff.lint.per-file-ignores]
+"earthaccess/search.py" = ["T20"]
```
Comment (botanical) [inline pyproject.toml:131]: I do not know if this is necessarily what we want but I didn't know how to ignore that we have intentional print statements 

> earthaccess/search.py:312:9: T201 `print` found
> earthaccess/search.py:313:9: T201 `print` found

File: tests/integration/test_auth.py (hunk 10 above)
Comment (mfisher87) [reply tests/integration/test_auth.py:96]: >In fact, I suspect the try/except should be removed.

>it seems intentional to me according to this commit 😯

It _does_ still seem weird. But let's be intentional about the scope of this PR :) Can we take this to an issue / discussion?

File: pyproject.toml (hunk 12 above)
Comment (mfisher87) [reply pyproject.toml:131]: On those lines, you can append a comment like `# noqa: T201` :) https://docs.astral.sh/ruff/linter/#error-suppression

**Review by chuckwondo — CHANGES_REQUESTED** [review]
@botanical, thanks for the cleanup of the `print` calls and adding the `ruff` rule. All looks good, except that I see numerous changes that have already landed on `main`, so it seems perhaps that your attempt to merge things from `main` went awry somewhere. For example, `.codespellignore` already exists on `main`, but shows up in this PR as an addition (among several other existing changes on `main`). Can you perhaps rebase your changes onto `main`? Happy to jump on a call to help sort it out if wrangling the commits gets gnarly.

File: tests/integration/test_auth.py (hunk 10 above)
Comment (chuckwondo) [reply tests/integration/test_auth.py:96]: > > In fact, I suspect the try/except should be removed.
> 
> > it seems intentional to me according to this commit 😯
> 
> It _does_ still seem weird. But let's be intentional about the scope of this PR :) Can we take this to an issue / discussion?

Sounds good to me. @botanical, would you mind opening a separate issue for this? Also, would you open yet another issue about refactoring loops in tests to instead be parametrized tests (`pytest.parametrize`, and yes, that is how it's spelled, which is apparently a valid variant of _parameterize_).

Comment (mfisher87) [reply tests/integration/test_auth.py:96]: >yes, that is how it's spelled, which is apparently a valid variant of parameterize

Gets me every time

Comment (botanical) [reply tests/integration/test_auth.py:96]: Issues added here:

- https://github.com/nsidc/earthaccess/issues/577
- https://github.com/nsidc/earthaccess/issues/578

File: CHANGELOG.md [hunk 13]
```diff
@@ -11,6 +11,8 @@
     intended use cases.
   * [#555](https://github.com/nsidc/earthaccess/issues/555): YAML formatting is
     now performed with `yamlfmt` instead of `prettier`.
+  * [#511](https://github.com/nsidc/earthaccess/issues/511): Replaced `print`
+    calls with `logging` calls, where appropriate and added T20 Ruff rule.
```
Comment (mfisher87) [inline CHANGELOG.md:15]:
```suggestion
    calls with `logging` calls where appropriate and added T20 Ruff rule.
```

**Review by mfisher87 — APPROVED** [review]
Thanks so much @botanical :star_struck: 

@chuckwondo (or anyone else) if you don't have any last thoughts I'll merge tomorrow!

**Comment by chuckwondo** [comment]
> Thanks so much @botanical 🤩
> 
> @chuckwondo (or anyone else) if you don't have any last thoughts I'll merge tomorrow!

@mfisher87, I'd like for us to use "Squash and merge" as a matter of habit. That's what I've been doing for PRs, as it keep the commit history much cleaner, unless there's an objection to that approach.

**Comment by jhkennedy** [comment]
> That's what I've been doing for PRs, as it keep the commit history much cleaner, unless there's an objection to that approach.

This could be a whole :fire:-war discussion :rofl: and it's come up on slack before, but for Earthaccess I think squash-merging matches the current development/release workflow the best.

I think we should disable all but 1 way to merge personally, so there's no question:
image

**Comment by mfisher87** [comment]
I think we can call this mergeable, and squash it is! Thanks again @botanical :bow: 

I don't have strong feelings about merge strategy, but I do feel strongly let's take it over here https://github.com/nsidc/earthaccess/issues/348 :)

**Pull request merged**